MapboxNavigation/NavigationMapView.swift

```diff
@@ -690,6 +690,45 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
         ].compactMap { style.source(withIdentifier: $0) }))
     }
     
+    /**
+     Removes completed sections of the route line.
+     */
+    
+    public func fadeRouteLine(percentComplete: Double) {
+        guard let style = self.style else { return }
+        
+        /**
+         This is necessary to prevent duplicate keys in the dictionary
+         defining the gradient. It's not for aesthetic purposes.
+         */
+        let gradientToTransparent: Double = 0.0005
+
+        var stops: [Double: UIColor] = [
+            0: self.trafficUnknownColor.withAlphaComponent(1),
+            1: self.trafficUnknownColor.withAlphaComponent(1)
+        ]
+
+        var casingStops: [Double: UIColor] = [
+            0: routeCasingColor.withAlphaComponent(1),
+            1: routeCasingColor.withAlphaComponent(1)
+        ]
+
+        if percentComplete > 0 && percentComplete + gradientToTransparent < 1 {
+            stops[0] = trafficUnknownColor.withAlphaComponent(0)
+            casingStops [0] = routeCasingColor.withAlphaComponent(0)
+            stops[percentComplete] = trafficUnknownColor.withAlphaComponent(0)
+            stops[percentComplete + gradientToTransparent] = trafficUnknownColor.withAlphaComponent(1)
+            casingStops[percentComplete] = routeCasingColor.withAlphaComponent(0)
+            casingStops[percentComplete + gradientToTransparent] = routeCasingColor.withAlphaComponent(1)
+        }
+
+        if let routeLine = style.layer(withIdentifier: StyleLayerIdentifier.route) as? MGLLineStyleLayer,
+           let routeLineCasing = style.layer(withIdentifier: StyleLayerIdentifier.routeCasing) as? MGLLineStyleLayer {
+            routeLine.lineGradient = NSExpression(format: "mgl_interpolate:withCurveType:parameters:stops:($lineProgress, 'linear', nil, %@)", stops)
```

Comment:
By default, `routeLineCasing` is a single LineString feature representing the entire leg, whereas `routeLine` is a FeatureCollection of LineStrings, one per traffic congestion segment. Applying the gradient to this layer causes the exotic dotted line pattern in the screenshot above. If we don’t apply the gradient to this layer, then there may be random-looking, disconnected yellow and red lines behind the puck.

Maybe we could split the difference by making the casing layer become 75% translucent, not transparent, behind the puck. That would preserve the erasing effect but ensure the user can intuit the purpose of those traffic congestion lines.

/cc @d-prukop @LukasPaczos

Reply:
On Android, we've converted the route line to be a single feature as well, to avoid the "dotted" appearance. The traffic is now represented with the percentage-based gradient expression instead of coloring single traffic segment features independently. As a side note, this also significantly reduced the memory footprint of the app because we don't have to split the route geometry into segments and store them.
/cc @cafesilencio

Reply:
On iOS we have a reference to both the entire route line and the individual segments. Both are rendered on the map, it seems. I'm not sure if the segments are used for anything else other than traffic coloring - @1ec5 do you know if we could get by with just rendering the single route (`routeLineCasing`) instead?

Reply:
The segments are only used for traffic congestion coloring. If we can implement them as part of an overall route line feature, then that would be ideal. I think we still need a second layer so that the fill is adequately cased – the style specification doesn’t provide many stroking options otherwise – but its shape source can contain a single feature instead of a multitude of them.

Reply:
> I think we still need a second layer so that the fill is adequately cased

Yes, that's also the case on Android. The route line and traffic are a single feature laid on top of route line casing (route line shield) which is another feature, equal to the same geometry but a little bit wider and with a different color. I missed mentioning that above.